### PR TITLE
Add hotreload

### PR DESCRIPTION
**Pull request by Evnsan**
Do not send messages to kafka if these messages are already there

### CODE REVIEW

File: kafka_topic_dumper/kafka_client.py [hunk 1]
```diff
@@ -185,73 +208,146 @@ def get_messages(self, num_messages_to_consume, max_package_size_in_msgs,
...
+        if beginning_offsets:
+            offsets, num_messages_available = self._calculate_offsets(
+                beginning_offsets=beginning_offsets,
+                end_offsets=end_offsets,
+                num_messages_to_consume=1)
+            self._set_offsets(offsets)
+            self.consumer.subscribe(self.dump_state_topic)
+            for message in self._get_messages(num_messages_available):
+                state = json.loads(message[1].decode())
+                if  state.get('topic_name', '') == self.topic and \
```
Comment (gustavo-momente) [inline kafka_topic_dumper/kafka_client.py]: extra space after `if`

File: kafka_topic_dumper/kafka_client.py [hunk 2]
```diff
@@ -185,73 +208,146 @@ def get_messages(self, num_messages_to_consume, max_package_size_in_msgs,
...
+        test_state = {
+            'test_id': test_id,
+            'topic_name': self.topic,
+            'offsets': test_offsets}
+
+        future = self.producer.send(
+            topic=self.dump_state_topic,
+            key=test_id,
+            value=json.dumps(test_state))
+        future.get(timeout=self.timeout)
```
Comment (gustavo-momente) [inline kafka_topic_dumper/kafka_client.py]: Isn't this timeout too low?

File: kafka_topic_dumper/kafka_client.py [hunk 3]
```diff
@@ -185,73 +208,146 @@ def get_messages(self, num_messages_to_consume, max_package_size_in_msgs,
...
+                                               Prefix=dump_path)
         file_names = []
         for response in response_iterator:
             if response['KeyCount'] > 0:
                 file_names = [(f['Key'], f['Size'])
                               for f in response['Contents']]
         file_names.sort()
+        return dump_prefix, file_names
+
+    def _set_test_topic(self, test_id):
```
Comment (gustavo-momente) [inline kafka_topic_dumper/kafka_client.py]: This name is confusing

File: kafka_topic_dumper/kafka_client.py [hunk 4]
```diff
@@ -185,73 +208,146 @@ def get_messages(self, num_messages_to_consume, max_package_size_in_msgs,
...
+            'offsets': test_offsets}
+
+        future = self.producer.send(
+            topic=self.dump_state_topic,
+            key=test_id,
+            value=json.dumps(test_state))
+        future.get(timeout=self.timeout)
+        logger.info('State saved')
+
+    def _get_test_state(self, test_id):
```
Comment (gustavo-momente) [inline kafka_topic_dumper/kafka_client.py]: This name is confusing

File: kafka_topic_dumper/kafka_client.py [hunk 5]
```diff
@@ -185,73 +208,146 @@ def get_messages(self, num_messages_to_consume, max_package_size_in_msgs,
...
+                logger.debug('File <{}> reloaded to kafka'.format(file_path))
+                remove(file_path)
+        else:
+            logger.info('Messages already uploaded. Just resetting offsets')
+            partitions = self._get_partitions(self.topic)
+            offsets = {}
+            for partition in partitions:
+                offsets[partition] = dump_offsets[str(partition.partition)]
+            self._set_offsets(offsets)
+            logger.debug('Should reset offsets to {}'.format(offsets))
```
Comment (gustavo-momente) [inline kafka_topic_dumper/kafka_client.py]: I think `Will reset offsets to {}`

File: kafka_topic_dumper/kafka_client.py [hunk 6]
```diff
@@ -185,73 +208,146 @@ def get_messages(self, num_messages_to_consume, max_package_size_in_msgs,
...
+                if  state.get('topic_name', '') == self.topic and \
+                    state.get('test_id', '') == test_id:
+                    return state.get('offsets', None)
+            return None
+
+    def reload_kafka_server(self, bucket_name, dir_path, dump_prefix=None):
+
+        s3_client = boto3.client('s3')
+
+        test_id, file_names = self._get_file_names(
```
Comment (gustavo-momente) [inline kafka_topic_dumper/kafka_client.py]: Do we really need to query s3 in the case where we'll only reset the offsets?

**Review by gustavo-momente — DISMISSED** [review]

File: kafka_topic_dumper/kafka_client.py (hunk 1 above)
Comment (Evnsan) [reply kafka_topic_dumper/kafka_client.py]: Fixed

File: kafka_topic_dumper/kafka_client.py (hunk 3 above)
Comment (Evnsan) [reply kafka_topic_dumper/kafka_client.py]: Fixed

File: kafka_topic_dumper/kafka_client.py (hunk 4 above)
Comment (Evnsan) [reply kafka_topic_dumper/kafka_client.py]: Fixed

File: kafka_topic_dumper/kafka_client.py (hunk 5 above)
Comment (Evnsan) [reply kafka_topic_dumper/kafka_client.py]: Fixed

File: kafka_topic_dumper/kafka_client.py (hunk 2 above)
Comment (Evnsan) [reply kafka_topic_dumper/kafka_client.py]: Not really, we are using this inside AWS. so it should not take so long.

File: kafka_topic_dumper/kafka_client.py (hunk 6 above)
Comment (Evnsan) [reply kafka_topic_dumper/kafka_client.py]: We need this to decide the test_id when we do not have a dump_prefix. We search the AWS-S3 bucket looking for the prefix.
But you're right, I can change it so I do not query AWS-S3 if dump_prefix was informed.

File: kafka_topic_dumper/main.py [hunk 7]
```diff
@@ -82,7 +88,7 @@ def main():
 
     bootstrap_servers = opts.bootstrap_servers
     bucket_name = opts.bucket_name
-    group_id = 'kafka_topic_dumper'
+    group_id = getattr(opts, 'reload_consumer_group', None)
```
Comment (danielkza) [inline kafka_topic_dumper/main.py]: Does this attribute need to be accessed here? Can't we access it normally inside the reload command condition?

File: kafka_topic_dumper/kafka_client.py [hunk 8]
```diff
@@ -269,6 +269,10 @@ def _get_file_names(self, bucket_name, dump_id):
                                   for f in response['Contents'])
         file_names.sort()
 
+        if len(file_names) == 0:
```
Comment (danielkza) [inline kafka_topic_dumper/kafka_client.py]: `if not file_names`

Comment (Evnsan) [reply kafka_topic_dumper/kafka_client.py]: Fixed

File: kafka_topic_dumper/main.py (hunk 7 above)
Comment (Evnsan) [reply kafka_topic_dumper/main.py]: We need this because we use it to create KafkaClient, not to call its methods.

**Review by danielkza — APPROVED** [review]

**Pull request merged**